### PR TITLE
WRKLDS-1449: bump(k8s): update k8s.io/* dependencies to v1.31.1

### PR DESCRIPTION
**Pull request by atiratree**
TODO:
- [x] depends on https://github.com/openshift/kubernetes-sigs-controller-tools/pull/20
- [x] verify-prerelease-lifecycle-gen.sh  should be fixed
- [x] not passing the verify-crd-schema-checker.sh: seeing a lot of `NoNewRequiredFields`
- [ ] OPTIONAL: the rest of the scripts should be ported to gengo: https://github.com/openshift/api/pull/1776

### CODE REVIEW

**Comment by atiratree** [comment]
The PR passes make update so we can use it as a dependency for now, but there are still TODOs which are mentioned in the description.

File: tools/codegen/pkg/openapi/generator.go [hunk 1]
```diff
@@ -24,7 +24,7 @@ var (
 		"k8s.io/apimachinery/pkg/runtime",
 		"k8s.io/apimachinery/pkg/util/intstr",
 		"k8s.io/apimachinery/pkg/api/resource",
-		"k8s.io/apimachinery/pkg/version",
+		"k8s.io/apimachinery/pkg/version/...", // Make version optional as it is not imported anywhere. It will be picked up if somebody starts using it in the future.
```
Comment (JoelSpeed) [inline tools/codegen/pkg/openapi/generator.go:27]: Do you know what error this was causing if we don't include this change?

File: openapi/generated_openapi/zz_generated.openapi.go [hunk 2]
```diff
@@ -9020,7 +9012,7 @@ func schema_openshift_api_config_v1_AuditCustomRule(ref common.ReferenceCallback
 						},
 					},
 				},
-				Required: []string{"group"},
+				Required: []string{"group", "profile"},
```
Comment (JoelSpeed) [inline openapi/generated_openapi/zz_generated.openapi.go:9015]: Newly required field?

File: openapi/generated_openapi/zz_generated.openapi.go [hunk 3]
```diff
@@ -20977,6 +20963,7 @@ func schema_openshift_api_config_v1alpha1_RetentionNumberConfig(ref common.Refer
 						},
 					},
 				},
+				Required: []string{"maxNumberOfBackups"},
```
Comment (JoelSpeed) [inline openapi/generated_openapi/zz_generated.openapi.go:20966]: Newly required field?

File: openapi/generated_openapi/zz_generated.openapi.go [hunk 4]
```diff
@@ -21047,6 +21034,7 @@ func schema_openshift_api_config_v1alpha1_RetentionSizeConfig(ref common.Referen
 						},
 					},
 				},
+				Required: []string{"maxSizeOfBackupsGb"},
```
Comment (JoelSpeed) [inline openapi/generated_openapi/zz_generated.openapi.go:21037]: Newly required field?

File: hack/update-prerelease-lifecycle-gen.sh [hunk 5]
```diff
@@ -8,4 +8,4 @@ if [ -z "${PRERELEASE_LIFECYCLE_GEN:-}" ];then
   PRERELEASE_LIFECYCLE_GEN="${TOOLS_OUTPUT}/prerelease-lifecycle-gen"
 fi
 
-"${PRERELEASE_LIFECYCLE_GEN}" --logtostderr -v 1 -h $(dirname "${BASH_SOURCE}")/boilerplate.go.txt --input-dirs ${API_PACKAGES} ${EXTRA_ARGS:-}
+"${PRERELEASE_LIFECYCLE_GEN}" --output-file zz_prerelease_lifecycle_generated.go --logtostderr -v 3 --go-header-file $(dirname "${BASH_SOURCE}")/boilerplate.go.txt ${EXTRA_ARGS:-} $(echo "${API_PACKAGES}" | tr ',' ' ')
```
Comment (JoelSpeed) [inline hack/update-prerelease-lifecycle-gen.sh:11]: Was this intentionally a permanent change?

File: machine/v1/tests/controlplanemachinesets.machine.openshift.io/stable.controlplanemachineset.aws.machineapimigration.testsuite.yaml [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+            spec:
+              providerSpec: {}
+              authoritativeAPI: MachineAPI
+            failureDomains:
+              platform: AWS
+              aws:
+              - subnet:
+                  type: Filters
+                  filters:
+                  - name: foo
```
Comment (JoelSpeed) [inline machine/v1/tests/controlplanemachinesets.machine.openshift.io/stable.controlplanemachineset.aws.machineapimigration.testsuite.yaml:170]: Nit: Missing final new line

File: machine/v1/tests/controlplanemachinesets.machine.openshift.io/stable.controlplanemachineset.aws.nomachineapimigration.testsuite.yaml [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+                machine.openshift.io/cluster-api-cluster: cluster
+            spec:
+              providerSpec: {}
+            failureDomains:
+              platform: AWS
+              aws:
+              - subnet:
+                  type: Filters
+                  filters:
+                  - name: foo
```
Comment (JoelSpeed) [inline machine/v1/tests/controlplanemachinesets.machine.openshift.io/stable.controlplanemachineset.aws.nomachineapimigration.testsuite.yaml:167]: Nit: Missing final new line

File: machine/v1/tests/controlplanemachinesets.machine.openshift.io/stable.controlplanemachineset.openstack.machineapimigration.testsuite.yaml [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+            spec:
+              providerSpec: {}
+              authoritativeAPI: MachineAPI
+            failureDomains:
+              platform: OpenStack
+              openstack:
+              - availabilityZone: foo
+                rootVolume:
+                  availabilityZone: foo
+                  volumeType: bar
```
Comment (JoelSpeed) [inline machine/v1/tests/controlplanemachinesets.machine.openshift.io/stable.controlplanemachineset.openstack.machineapimigration.testsuite.yaml:320]: Nit: Missing final new line

File: machine/v1/tests/controlplanemachinesets.machine.openshift.io/stable.controlplanemachineset.openstack.nomachineapimigration.testsuite.yaml [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+                machine.openshift.io/cluster-api-cluster: cluster
+            spec:
+              providerSpec: {}
+            failureDomains:
+              platform: OpenStack
+              openstack:
+              - availabilityZone: foo
+                rootVolume:
+                  availabilityZone: foo
+                  volumeType: bar
```
Comment (JoelSpeed) [inline machine/v1/tests/controlplanemachinesets.machine.openshift.io/stable.controlplanemachineset.openstack.nomachineapimigration.testsuite.yaml:314]: Nit: Missing final new line

File: hack/verify-prerelease-lifecycle-gen.sh [hunk 10]
```diff
@@ -2,4 +2,12 @@
 
 source "$(dirname "${BASH_SOURCE}")/lib/init.sh"
 
-EXTRA_ARGS=--verify-only ${SCRIPT_ROOT}/hack/update-prerelease-lifecycle-gen.sh
+${SCRIPT_ROOT}/hack/update-prerelease-lifecycle-gen.sh
```
Comment (JoelSpeed) [inline hack/verify-prerelease-lifecycle-gen.sh:5]: Verify only no longer works?

File: operator/v1/types_network.go [hunk 11]
```diff
@@ -9,6 +9,7 @@ import (
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 // +kubebuilder:object:root=true
 // +kubebuilder:resource:path=networks,scope=Cluster
+// +kubebuilder:subresource:status
```
Comment (JoelSpeed) [inline operator/v1/types_network.go:12]: What are the implications for the network operator of this change? Have we tested upgrades to make sure this doesn't break?

**Review by JoelSpeed — COMMENTED** [review]
There appear to be a bunch of new required fields picked up in the openapi schema, but I didn't see these in the CRD schemas, did you look into this at all?

File: machine/v1/tests/controlplanemachinesets.machine.openshift.io/stable.controlplanemachineset.aws.machineapimigration.testsuite.yaml (hunk 6 above)
Comment (atiratree) [reply machine/v1/tests/controlplanemachinesets.machine.openshift.io/stable.controlplanemachineset.aws.machineapimigration.testsuite.yaml:170]: thanks, fixed

File: hack/verify-prerelease-lifecycle-gen.sh (hunk 10 above)
Comment (atiratree) [reply hack/verify-prerelease-lifecycle-gen.sh:5]: yeah, unfortunately this option was removed from code-generator/gengo

File: operator/v1/types_network.go (hunk 11 above)
Comment (atiratree) [reply operator/v1/types_network.go:12]: The newly generated CRD was imported and tested in https://github.com/openshift/cluster-network-operator/pull/2506. The upgrade path seems to work fine.

Comment (JoelSpeed) [reply operator/v1/types_network.go:12]: Looking at that job, there are three upgrade from 4.17 jobs that all failed indicating CSI issues, are we sure this isn't the cause of that?

Comment (atiratree) [reply operator/v1/types_network.go:12]: the status update problems observed in https://github.com/openshift/cluster-network-operator/pull/2506#issuecomment-2383308291 should be fixed with https://github.com/openshift/cluster-network-operator/pull/2506/commits/febd41376cc3553fa47a3be8cb9503337a80934d

**Comment by atiratree** [comment]
verify [failed](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_api/1983/pull-ci-openshift-api-master-verify/1841070314142830592) because we need to react to https://github.com/openshift/api/pull/2027/files that merged in the meantime and regenerate the CRDs

**Pull request merged**